### DISCUSSION TITLE
Welcome to jailmaker Discussions!

### DISCUSSION THREAD

**Discussion by Jip-Hop** [Jip-Hop/jailmaker#19] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by spusuf** [comment]
Hey, I'm looking to switch from Core to Scale and jails are the critical feature I'd miss most, I hate that installing a VM to run docker has become the norm. Tried this jailmaker script and installed docker and put simply I've never had a simpler docker install and setup process EVER. Love seeing my cpu utilisation say "12 cores at 0%", instead of 10-20% usage while doing nothing. Hope to contribute and bug test.

**Comment by neilkpd** [comment]
Hello.
I would like to begin testing.
Should I use the current stable, TrueNAS SCALE 22.12.3.3?
I'm currently using the current master, TrueNAS-SCALE-23.10-MASTER-20230811-024711 as Admin user
and got stuck using the Docker convience script at
sh install-docker.sh --dry-run
error message: 502: dpkg: Permission denied

**Reply by spusuf** [reply]
try sudo

**Reply by spusuf** [reply]
Yes 22.13.3 is tested known working, won't really be able to give too much help on nightlies but it sounds like youve got the jails set up and youre running into docker issues. Try sudo as thats likely your problem and I'll test 23.10 when the beta comes out in a couple days.

**Reply by spusuf** [reply]
Try using the method linked here to enable dpkg, it might just be disabled (default TrueNAS behaviour) and the docker installer is trying to use it. You might need to do it inside the jail, you might need to do it in the TrueNAS shell. 

https://github.com/Jip-Hop/jailmaker/issues/16#issuecomment-1592401680

**Reply by Jip-Hop** [reply]
Testing on anything beyond 22.12.3.3 (including nightlies) is very welcome. I haven't tried anything beyond 22.12.3.3 myself.

**Comment by neilkpd** [comment]
Using sudo gave the same error. 502: dpkg: Permission denied

**Reply by Jip-Hop** [reply]
Did you successfully create your first jail? And did you [open a shell inside the jail](https://github.com/Jip-Hop/jailmaker#jail-shell)? If your goal is to install docker inside the jail (which is what I use jailmaker for personally), then be sure to run the install-docker.sh script from a shell inside the jail (and not from the shell on the TrueNAS host). You should not get dpkg permission errors inside the jail.

**Comment by neilkpd** [comment]
I followed @eramdam and re-enabled apt and dpkg (chmod +x /bin/apt*, chmod +x /bin/dpkg*)
Docker is now installed.
Now I need to install Portainer.

**Reply by Jip-Hop** [reply]
This doesn't sound right... I don't know what you try to do but with `jlmkr.py install` you now no longer need to manually enable apt and dpkg and they definitely don't have anything to do with installing docker. I suggest you make a dedicated question topic in the Discussions section if you need help (instead of this welcome thread) and post more information on your goals and the steps you're taking together with screenshots/logging.

**Comment by neilkpd** [comment]
systemd-nspawn comes pre-installed in
TrueNAS-SCALE-23.10-MASTER-20230811-024711

Portainer is installed, but I have to research not having to use sudo all the time for each application

**Comment by Kieros** [comment]
I realy like this setup. I have been struggling from the beginning with truenas scale and the ix app system. Not sure how they are heading, running docker in docker, K3s kubernetes, helm, rancher on the background.  It is very confusing. I have a few containers setup with the launch image thing. But it has flaws. Can't pass through zwave usb for example. Running this natively with jail feels good and like a straight direction. Tried running this before but it broke network in the vm's with a docker script. As I run pfsense in a vm the internet stopped working. 

I wonder, as I am pretty new to linux, there is a steep learning curve. But hanging in there. How would you guys set this up as experienced users? Directory wise etc. I have bound a data directory with using the bind method. But I have trouble for example with portainer to put containers or stacks in this data folder. When I start this I would like to organise this properly from the start. Any advice would be welcome. From you mystro's.

Also how is this realy efficiënt? Running debian in jail and then docker engine on top of that? How are resources devided is debian the host for docker within the jail having fixed memory, cpu? To run these containers of? Or are the containers getting the resources directly from truenas host or the debian host? A little bit confused when talking about the host. 

Sorry if to many questions. But keep up the good work. I think I will start testing with debian 12

**Reply by spusuf** [reply]
If you mean pfSense, you really shouldnt have it running in a VM on truenas (proxmox is more robust hypervisor, and pfsense recommends bare metal). But if you understand the security implications and the fact it may break multiple things (like youve experienced) then go for it. I'm using macvlan on all my jails which I can set to a static IP, very similar to how networking functions in truenas core. The default networking uses your host network, so if youve changed something to get your VM working then youd have to describe your setup for anyone to help.

**Reply by spusuf** [reply]
I agree the current TrueNAS apps solution isnt perfect and systemd containers seem to be much better in every way.

As for setting up directories, when you create a jail the image will create a linux filesystem for you. From there, most apps will have a default linux location for their files whether `/root`, `/home` ,or `/media`. For example a media application like Jellyfin, Plex, or Immich might put its files in `/media` from which you can bind to either the /media of the host (truenas) which is bad, dont do it. or to a dataset inside a pool (e.g `/mnt/pool/mymedia`).  With portainer or any docker app check if you can manually set a volume in the compose file, if not then it may be hard coded into the image. When you find where its putting files, just bind the folders you need regular access to.

For my truenas directory ive got a jails dataset which contains jailmaker (including the script and the actual jails and their filesystems), a backup folder (for manually backing up jails), and a scripts folder (with scripts for installing core software and docker).

As for how its MUCH more efficient than a VM of Linux, its not actually running a whole separate kernel, its running on the host kernel (the linux install thats underneath truenas). That way you dont have to give the VM any virtual CPUs or memory, its running on the truenas resources. Anything TrueNAS has access to (cpu and memory) the jails will have access to and use as needed. There is the possibility of the jails using so much resources that the other functions become laggy and im sure systemd has an allocation limit somewhere, but if you manage your system properly it shouldnt use more than necessary. Theres no kubernetes overhead (10% cpu usage with no apps installed for me) like the truenas apps, or no virtual resources like a VM. From my understanding its running on the same linux install that truenas is but with separated storage and system calls for security purposes (and luckily for us means it wont touch or corrupt a truenas install).

**Reply by Kieros** [reply]
Thank you for the explanation. When I get back home I will  show how the vm is setup with truenas and pfsense.
I started off with proxmox. But then I tought it would be better to run truenas on bare metal when thinking about resources. Are you running truenas on proxmox? 

Should it not be possible to setup pfsense in a way like this?  Next to truenas or at the same level like this jail?
So not running in proxmox? 

If running everything in proxmox, then using this jail is not the way I think? Then proxmox can handle the docker native thing in a better way?

**Comment by Kieros** [comment]
About the vm in truenas scale.
I have the following setup.
2 NICS
At truenas scale network
NIC1 - enp5s0 - "ISP" ticked DHCP and Autoconfigure IPv6 MTU1500
NIC2 - enp10s0 - "network" DHCP and Autoconfigure IPv6 Not ticked MTU1500

Both nics bridged. 
br0 - member: enp10s0 - MTU1500
br1 - member: enp5s0 -MTU1500 < bridge necessary for VM to recognize KPN it won't get an IP if I use the enp5s0

VM settings
Devices:
2 NICS both VirtIO
NIC1 attached: br1- ISP
NIC2 attached: br0 - network
![image](https://github.com/Jip-Hop/jailmaker/assets/99768682/cb480512-a871-4b4e-b145-cb7bb7a32775)

In pfsense VM
I have vtnet0 and vtnet1
vtnet0 is connected with br1
vtnet1 is connected with br0

vtnet 0 is setup to vlan 6 and vlan 4
vlan 6 is KPN_internet (pppoe0)
vlan 4 is KPN_iTV (not used at all) as I do not have tv.

btw the only trouble I had was due to this docker script I used before which messed up the routing of the vm.
Like I read on the other Git as well. It has been running for a year without problems.
But how is this insecure?
